dids/diddht/diddht.go

```diff
@@ -1,175 +1,191 @@
+type dHTDidOptions struct {
+	algorithmID         string
+	keyManager          crypto.KeyManager
+	services            []*didcore.Service
+	verificationMethods []didcore.VerificationMethod
+	purposes            map[string][]didcore.Purpose
+	relay               relay
+}
 
-			if !ok {
-				// no relationships
-				continue
-			}
+type DHTDidOption func(o *dHTDidOptions)
 
-			opts := []didcore.Purpose{}
-			for _, r := range relationships {
-				if o, ok := vmPurposeDNStoDID[r]; ok {
-					opts = append(opts, didcore.Purpose(o))
-				}
-			}
+// WithServices adds the provided services to the DID document.
+func WithServices(services ...*didcore.Service) DHTDidOption {
+	return func(o *dHTDidOptions) {
+		o.services = append(o.services, services...)
+	}
+}
 
-			document.AddVerificationMethod(
-				vMethod,
-				didcore.Purposes(opts...),
-			)
-		case strings.HasPrefix(name, "_s"):
-			var s didcore.Service
-			if err := UnmarshalService(data, &s); err != nil {
-				// TODO handle error
-				continue
+// WithVerificationMethod adds the provided verification method to the DID document.
+func WithVerificationMethod(method didcore.VerificationMethod, purposes []didcore.Purpose) DHTDidOption {
+	return func(o *dHTDidOptions) {
+		o.verificationMethods = append(o.verificationMethods, method)
+		for _, p := range purposes {
+			if _, ok := o.purposes[method.ID]; !ok {
+				o.purposes[method.ID] = []didcore.Purpose{}
 			}
-			document.AddService(&s)
-		case strings.HasPrefix(name, "_cnt"):
-			// TODO add controller https://did-dht.com/#controller
-			// optional field
-			// comma-separated list of controller DID identifiers
-			document.Controller = strings.Split(data, ",")
-		case strings.HasPrefix(name, "_aka"):
-			// TODO add aka https://did-dht.com/#also-known-as
-			document.AlsoKnownAs = strings.Split(data, ",")
-		default:
+			o.purposes[method.ID] = append(o.purposes[method.ID], p)
 		}
 	}
+}
 
-	return document, nil
+// WithKeyManager sets the key manager to use for generating the DID's keypair.
+func WithKeyManager(k crypto.KeyManager) DHTDidOption {
+	return func(o *dHTDidOptions) {
+		o.keyManager = k
+	}
 }
 
-// parseDNSDID takes the bytes of the DNS representation of a DID and creates an internal representation
-// used to create a DID document
-func parseDNSDID(data []byte) (*Decoder, error) {
-	var p dnsmessage.Parser
-	if _, err := p.Start(data); err != nil {
-		return nil, err
+// WithRelay sets the relay to use for publishing the DID to the DHT.
+func WithRelay(relayURL string, client *http.Client) DHTDidOption {
+	return func(o *dHTDidOptions) {
+		o.relay = pkarr.NewPkarrRelay(relayURL, client)
 	}
+}
+
+// Create creates a new `did:dht` DID and publishes it to the DHT network via a Pkarr relay.
+//
+// If no relay is passed in the options, Create uses a default Pkarr relay.
+// Spec: https://did-dht.com/#create
+func Create(opts ...DHTDidOption) (*did.BearerDID, error) {
```

Comment:
any specific rationale behind returning a pointer here? curious if it's more for stylistic reasons or if the motivation is performance based. `didjwk.Create` doesn't return a pointer. hoping for pointer usage to be consistent across the codebase so that it's easy to reason about

Reply:
Performance-wise - it is likely more efficient to not make copies of structs. There are nuances to this, depending on the struct size. To avoid thinking about those nuances all the time, we can flip this to copying a struct _sometimes_ is better. So defaulting to using pointers for structs is the better option most of the time.
Nil - returning `struct, nil` or `nil, err` is a very common pattern in Go, as opposed to returning a copy of an empty struct. 

These are some of my main points in defaulting to pointers when returning a newly created object. I empathize with the consistency point, though from a callers perspective, I think it is largely the same code whether a `value, error` or `pointer, error` is returned.

To make this actionable, the options I'm thinking is:
1. change this to return a value, to keep the repo consistent and if there are no objections, in a separate PR change to returning pointers here and in the rest of the repo
2. keep it, because the calling code shouldn't be affected, and make a separate pr to update the rest of the repo (assuming there are no objections to returning pointers from factory functions)
3. Update the rest of the repo in this PR (I don't think this is a real option... but I'm open to feedback)

Any preference?

Reply:
> change this to return a value, to keep the repo consistent and if there are no objections, in a separate PR change to returning pointers here and in the rest of the repo

This one is my vote b/c I think we should come to further consensus... I'm still suspect of pointers as being better